modules/dop/providers/projectpipeline/service_impl.go

```diff
@@ -1351,3 +1359,92 @@ func (p *ProjectPipelineService) ListUsedRefs(ctx context.Context, params deftyp
 	}
 	return resp.Ref, nil
 }
+
+type pipelineCategoryRule struct {
+	Key        string
+	Category   string
+	Rules      []string
+	RunningNum uint64
+	FailedNum  uint64
+	TotalNum   uint64
+}
+
+func (p *ProjectPipelineService) ListPipelineCategoryRule(ctx context.Context) []pipelineCategoryRule {
+	return []pipelineCategoryRule{
+		{
+			Key:      "build-deploy",
+			Category: p.trans.Text(apis.Language(ctx), "BuildDeploy"),
+			Rules:    []string{"pipeline.yml"},
+		},
+		{
+			Key:      "build-artifact",
+			Category: p.trans.Text(apis.Language(ctx), "BuildArtifact"),
+			Rules:    []string{".erda/pipelines/ci-artifact.yml", ".dice/pipelines/ci-artifact.yml"},
+		},
+		{
+			Key:      "others",
+			Category: p.trans.Text(apis.Language(ctx), "Uncategorized"),
+			Rules:    nil,
+		},
+	}
+}
+
+func (p *ProjectPipelineService) ListPipelineCategory(ctx context.Context, params *pb.ListPipelineCategoryRequest) (*pb.ListPipelineCategoryResponse, error) {
+	if err := params.Validate(); err != nil {
+		return nil, apierrors.ErrListProjectPipelineCategory.InvalidParameter(err)
+	}
+	project, err := p.bundle.GetProject(params.ProjectID)
+	if err != nil {
+		return nil, apierrors.ErrListProjectPipelineCategory.InternalError(err)
+	}
+
+	org, err := p.bundle.GetOrg(project.OrgID)
+	if err != nil {
+		return nil, apierrors.ErrListProjectPipelineCategory.InternalError(err)
+	}
+
+	staticsResp, err := p.PipelineDefinition.StatisticsGroupByFilePath(ctx, &dpb.PipelineDefinitionStatisticsRequest{
+		Location: makeLocation(&apistructs.ApplicationDTO{
+			OrgName:     org.Name,
+			ProjectName: project.Name,
+		}, cicdPipelineType),
+	})
+	if err != nil {
+		return nil, apierrors.ErrListProjectPipelineCategory.InternalError(err)
+	}
+	categoryRules := p.ListPipelineCategoryRule(ctx)
+
+	for _, statics := range staticsResp.PipelineDefinitionStatistics {
```

Comment:
too complex, encapsulate a func like: `sumByCategory(key, allStatistics)`, and then
```
for i := range categories {
  r = rule[i]
  r.PipelineStatisticsNums = sumByCategory(r.key, defStatistics)
  ...
}
```

Reply:
done